sympy/solvers/ode/single.py

```diff
@@ -425,42 +432,43 @@ def _get_general_solution(self, *, simplify: bool = True):
         return [gensol]
 
 
-
 class AlmostLinear(SinglePatternODESolver):
     r"""
     Solves an almost-linear differential equation.
 
     The general form of an almost linear differential equation is
 
-    .. math:: f(x) g(y) y + k(x) l(y) + m(x) = 0
-                \text{where} l'(y) = g(y)\text{.}
+    .. math:: a(x) g'(f(x)) f'(x) + b(x) g(f(x)) + c(x)
 
-    This can be solved by substituting `l(y) = u(y)`.  Making the given
-    substitution reduces it to a linear differential equation of the form `u'
-    + P(x) u + Q(x) = 0`.
+    This can be solved by substituting `g(f(x)) = u`.  Making the given
+    substitution reduces it to a linear differential equation of the form
+    `a(x) u' + b(x) u + c(x) = 0` which can now be solved for `f(x)`.
 
     The general solution is
 
         >>> from sympy import Function, dsolve, Eq, pprint
         >>> from sympy.abc import x, y, n
-        >>> f, g, k, l = map(Function, ['f', 'g', 'k', 'l'])
-        >>> genform = Eq(f(x)*(l(y).diff(y)) + k(x)*l(y) + g(x), 0)
+        >>> a, b, c, u = map(Function, ['a', 'b', 'c', 'u'])
+        >>> genform = Eq(a(x)*(u(x).diff(x)) + b(x)*u(x) + c(x), 0)
```

Comment:
This isn't really the general form. The general form is
```
In [9]: a, b, c, f, g = symbols('a, b, c, f, g', cls=Function)                                                                                 

In [10]: genform = a(x)*g(f(x)).diff(f(x))*f(x).diff(x) + b(x)*g(f(x)) + c(x)                                                                  

In [11]: genform                                                                                                                               
Out[11]: 
     d          d                                 
a(x)⋅──(f(x))⋅─────(g(f(x))) + b(x)⋅g(f(x)) + c(x)
     dx       df(x) 
```
This can't be solved though without knowing how to invert `g` so we need to substitute a function for `g` to get a particular example e.g.:
```

In [17]: example = genform.subs(g, sin).doit()                                                                                                 

In [18]: example                                                                                                                               
Out[18]: 
               d                               
a(x)⋅cos(f(x))⋅──(f(x)) + b(x)⋅sin(f(x)) + c(x)
               dx   
```
In principle that should be solvable but it leads to slow integrals...

Maybe setting a, b and c to 1 would help:
```
In [31]: example = cos(f(x))*f(x).diff(x) + sin(f(x)) + 1                                                                                      

In [32]: example                                                                                                                               
Out[32]: 
                      d           
sin(f(x)) + cos(f(x))⋅──(f(x)) + 1
                      dx          

In [33]: dsolve(example, f(x), hint='almost_linear')                                                                                           
Out[33]: 
⎡               ⎛    -x    ⎞             ⎛    -x    ⎞⎤
⎣f(x) = π - asin⎝C₁⋅ℯ   - 1⎠, f(x) = asin⎝C₁⋅ℯ   - 1⎠⎦
```

Reply:
> This isn't really the general form. The general form is
> 
> ```
> In [9]: a, b, c, f, g = symbols('a, b, c, f, g', cls=Function)                                                                                 
> 
> In [10]: genform = a(x)*g(f(x)).diff(f(x))*f(x).diff(x) + b(x)*g(f(x)) + c(x)                                                                  
> 
> In [11]: genform                                                                                                                               
> Out[11]: 
>      d          d                                 
> a(x)⋅──(f(x))⋅─────(g(f(x))) + b(x)⋅g(f(x)) + c(x)
>      dx       df(x) 
> ```

I think they are the same thing since this is just an expanded version of `(u(x).diff(x))` using chain rule.

Reply:
Yes but the ODE should be in terms of `f` which is the dependent variable.

Reply:
So do you suggest to change the example doctest and rewrite it in terms of f(x) with simplified coefficient functions and `g`?

Reply:
It should be changed because right now the example is not really the general form and is a linear ODE. The point of this method is that the substitution turns a nonlinear ODE into a linear one so the example should show solving a nonlinear ODE.